### ISSUE TITLE
User journal corruption on btrfs

### ISSUE THREAD

**Issue by mrc0mmand** [systemd/systemd#31222]
This is something I've been seeing in my _VM farm_ for a while and never have gotten into debugging it properly until now.

I use Vagrant a lot for my test VMs - the important bits are that `vagrant ssh` will log you into the VM as a `vagrant` user and I usually follow this with `sudo su -` to get a root shell.

While reproducing some issues involving grepping journal (like https://github.com/systemd/systemd/issues/30886), I quite often noticed `journalctl` complaining about corrupted journals. After collecting some data I found out that this seems to affect only VMs that run btrfs (Arch, Fedora), it's always the user journal, and it _always_ involves the _unpriv_user_ -> _root_ "session chain". Invoking `journalctl --rotate` from the nested root session is enough to corrupt the just archived journal.

I managed to minimize it into a very simple reproducer using our test suite:
```diff
diff --git a/test/TEST-04-JOURNAL/test.sh b/test/TEST-04-JOURNAL/test.sh
index 42574ce8ea..db7b4bc6a8 100755
--- a/test/TEST-04-JOURNAL/test.sh
+++ b/test/TEST-04-JOURNAL/test.sh
@@ -27,6 +27,10 @@ test_append_files() {
         mkdir -p "$dropin_dir"
         printf '[Service]\nStandardOutput=journal+console\nStandardError=journal+console' >"$dropin_dir/99-stdout.conf"
     fi
+
+    image_install sudo
+    inst /etc/sudoers
+    inst_recursive /usr/libexec/sudo/
 }
 
 do_test "$@"
diff --git a/test/test-functions b/test/test-functions
index 80fdcd26b9..cdae1e97b0 100644
--- a/test/test-functions
+++ b/test/test-functions
@@ -2099,7 +2099,7 @@ install_testuser() {
     # create unprivileged user for user manager tests
     mkdir -p "${initdir:?}/etc/sysusers.d"
     cat >"$initdir/etc/sysusers.d/testuser.conf" <<EOF
-u testuser    4711     "Test User" /home/testuser
+u testuser    4711     "Test User" /home/testuser /bin/bash
 EOF
 
     mkdir -p "$initdir/home/testuser"
diff --git a/test/units/testsuite-04.user-corrupt.sh b/test/units/testsuite-04.user-corrupt.sh
new file mode 100755
index 0000000000..74453497f1
--- /dev/null
+++ b/test/units/testsuite-04.user-corrupt.sh
@@ -0,0 +1,31 @@
+#!/usr/bin/env bash
+# SPDX-License-Identifier: LGPL-2.1-or-later
+set -eux
+set -o pipefail
+
+mkdir -p /etc/sudoers.d/
+echo "testuser ALL=(ALL) NOPASSWD: ALL" >/etc/sudoers.d/testuser
+
+REPRO_SCRIPT="$(mktemp)"
+
+cat >"$REPRO_SCRIPT" <<\EOF
+#!/bin/bash
+set -eux
+set -o pipefail
+
+journalctl --rotate --vacuum-files=1
+# Nuke all archived journals, so we start with a clean slate
+rm -f /var/log/journal/$(</etc/machine-id)/system@*.journal
+rm -f /var/log/journal/$(</etc/machine-id)/user-*@*.journal
+journalctl --header | grep path
+
+for i in {0..10}; do
+    journalctl --rotate
+    journalctl --sync
+    SYSTEMD_LOG_LEVEL=debug journalctl -n1 -q
+    (! journalctl -n0 -q |& grep corrupted)
+done
+EOF
+chmod +x "$REPRO_SCRIPT"
+
+systemd-run --wait --pipe --user -M testuser@ -- sudo "$REPRO_SCRIPT"
```

```
# make -C test/TEST-04-JOURNAL/ clean setup run TEST_MATCH_SUBTEST=user-corrupt TEST_NO_QEMU=1 BUILD_DIR=$PWD/build KERNEL_APPEND=systemd.default_standard_output=journal+console FSTYPE=btrfs
...
[945170.307432] testsuite-04.sh[97]: + journalctl --header
[945170.307748] testsuite-04.sh[98]: + grep path
[945170.311593] testsuite-04.sh[98]: File path: /var/log/journal/9059aaf50f1741469b471cee9e6056be/system.journal
[945170.311593] testsuite-04.sh[98]: File path: /var/log/journal/9059aaf50f1741469b471cee9e6056be/user-4711.journal
[945170.311849] testsuite-04.sh[91]: + for i in {0..10}
[945170.311926] testsuite-04.sh[91]: + journalctl --rotate
[945170.444078] testsuite-04.sh[91]: + journalctl --sync
[945170.449523] testsuite-04.sh[91]: + SYSTEMD_LOG_LEVEL=debug
[945170.449523] testsuite-04.sh[91]: + journalctl -n1 -q
[945170.471060] testsuite-04.sh[105]: Considering root directory '/run/log/journal'.
[945170.471060] testsuite-04.sh[105]: Root directory /run/log/journal added.
[945170.471060] testsuite-04.sh[105]: Considering root directory '/var/log/journal'.
[945170.471060] testsuite-04.sh[105]: Root directory /var/log/journal added.
[945170.471060] testsuite-04.sh[105]: Considering directory '/var/log/journal/9059aaf50f1741469b471cee9e6056be'.
[945170.471060] testsuite-04.sh[105]: Directory /var/log/journal/9059aaf50f1741469b471cee9e6056be added.
[945170.471060] testsuite-04.sh[105]: Journal effective settings seal=no keyed_hash=yes compress=ZSTD compress_threshold_bytes=8B
[945170.471060] testsuite-04.sh[105]: File /var/log/journal/9059aaf50f1741469b471cee9e6056be/system.journal added.
[945170.471060] testsuite-04.sh[105]: File /var/log/journal/9059aaf50f1741469b471cee9e6056be/system@23f5950dc21b4fd4be3d5391ec94d0e1-0000000000000a8b-000610b8052a115a.journal added.
[945170.471060] testsuite-04.sh[105]: File /var/log/journal/9059aaf50f1741469b471cee9e6056be/user-4711.journal added.
[945170.471060] testsuite-04.sh[105]: Failed to open journal file /var/log/journal/9059aaf50f1741469b471cee9e6056be/user-4711@23f5950dc21b4fd4be3d5391ec94d0e1-0000000000000000-0000000000000000.journal: Bad message
[945170.471060] testsuite-04.sh[105]: Journal file /var/log/journal/9059aaf50f1741469b471cee9e6056be/user-4711@23f5950dc21b4fd4be3d5391ec94d0e1-0000000000000000-0000000000000000.journal corrupted, ignoring file.
[945170.471060] testsuite-04.sh[105]: Journal filter: none
[945170.471060] testsuite-04.sh[105]: Root directory /var/log/journal removed.
[945170.471060] testsuite-04.sh[105]: Root directory /run/log/journal removed.
[945170.471060] testsuite-04.sh[105]: Directory /var/log/journal/9059aaf50f1741469b471cee9e6056be removed.
[945170.471060] testsuite-04.sh[105]: mmap cache statistics: 234 category cache hit, 8 window list hit, 4 miss
[945170.471060] testsuite-04.sh[105]: Feb 06 16:13:27 H testsuite-04.sh[91]: + journalctl --sync
[945170.494490] testsuite-04.sh[107]: + journalctl -n0 -q
[945170.494749] (sd-pam)[77]: pam_systemd(login:session): Failed to release session: Access denied
[945170.494841] testsuite-04.sh[108]: + grep corrupted
[945170.494841] testsuite-04.sh[108]: Journal file /var/log/journal/9059aaf50f1741469b471cee9e6056be/user-4711@23f5950dc21b4fd4be3d5391ec94d0e1-0000000000000000-0000000000000000.journal corrupted, ignoring file.
[945170.495458] testsuite-04.sh[60]: Finished with result: exit-code
[945170.495458] testsuite-04.sh[60]: Main processes terminated with: code=exited/status=1
[945170.495458] testsuite-04.sh[60]: Service runtime: 439ms
[945170.495458] testsuite-04.sh[60]: CPU time consumed: 9ms
[945170.495458] testsuite-04.sh[60]: Memory peak: 2.9M (swap: 0B)
[945170.496155] testsuite-04.sh[51]: + echo 'Subtest /usr/lib/systemd/tests/testdata/units/testsuite-04.user-corrupt.sh failed'
[945170.496155] testsuite-04.sh[51]: Subtest /usr/lib/systemd/tests/testdata/units/testsuite-04.user-corrupt.sh failed
[945170.496155] testsuite-04.sh[51]: + return 1
[945170.496520] systemd[69]: run-u0.service: Failed with result 'exit-code'.
[945170.499946] dbus-broker[64]: A security policy denied :1.15 to send method call /org/freedesktop/login1:org.freedesktop.login1.Manager.ReleaseSession to org.freedesktop.login1.
[945170.500688] systemd[1]: testsuite-04.service: Failed with result 'exit-code'.
[945170.500892] systemd[1]: Failed to start testsuite-04.service.
```

Here's one example journal from Fedora Rawhide with the latest main (ATTOW): https://mrc0mmand.fedorapeople.org/user-1000-rawhide-btrfs-corrupted.journal

As mentioned above, this seems to happen only with btrfs - xfs or ext4 (cc @bluca ;)) seem to work fine.

**Comment by bluca** [comment]
> As mentioned above, this seems to happen only with btrfs - xfs or ext4 (cc @bluca ;)) seem to work fine.

Close+wontfix as user error

**Comment by yuwata** [comment]
@mrc0mmand Could you test #31248 ?

**Comment by yuwata** [comment]
Ugh...

**Comment by yuwata** [comment]
Interestingly, 0000000000000000-0000000000000000.journal is not created on non-btrfs case.

By disabling `deferred_closes`, 0000000000000000-0000000000000000.journal is not created even on btrfs and fix the issue.
```diff
diff --git a/src/journal/journald-server.c b/src/journal/journald-server.c
index b7f7f5aee8..8fa3be3865 100644
--- a/src/journal/journald-server.c
+++ b/src/journal/journald-server.c
@@ -540,7 +540,7 @@ static int server_do_rotate(
                 (seal ? JOURNAL_SEAL : 0) |
                 JOURNAL_STRICT_ORDER;
 
-        r = journal_file_rotate(f, s->mmap, file_flags, s->compress.threshold_bytes, s->deferred_closes);
+        r = journal_file_rotate(f, s->mmap, file_flags, s->compress.threshold_bytes, /*s->deferred_closes*/NULL);
         if (r < 0) {
                 if (*f)
                         return log_ratelimit_error_errno(r, JOURNAL_LOG_RATELIMIT,
```
Also, passing filename rather than fd based path to copy_file_atomic() can fix the issue, and 0000000000000000-0000000000000000.journal neither created:
```diff
diff --git a/src/shared/journal-file-util.c b/src/shared/journal-file-util.c
index 8b41f5adc3..a7ac48f144 100644
--- a/src/shared/journal-file-util.c
+++ b/src/shared/journal-file-util.c
@@ -210,7 +210,7 @@ static void journal_file_set_offline_internal(JournalFile *f) {
 
                                 log_debug_errno(r, "Failed to re-enable copy-on-write for %s: %m, rewriting file", f->path);
 
-                                r = copy_file_atomic_full(FORMAT_PROC_FD_PATH(f->fd), f->path, f->mode,
+                                r = copy_file_atomic_full(f->path, f->path, f->mode,
                                                           0,
                                                           FS_NOCOW_FL,
                                                           COPY_REPLACE | COPY_FSYNC | COPY_HOLES | COPY_ALL_XATTRS,
```
But, I have no idea why these patch fix the issue. Will try to tackle the issue further.

**Comment by yuwata** [comment]
Random thoughts.

If a journal file contains no journal entry, I think we should not create 0000000000000000-0000000000000000.journal. If it is correctly created, the next rotation may fail as journald try to create 0000000000000000-0000000000000000.journal again.

Maybe, we should use something like `asynchronous_close()` rather than using threads for archiving journal files.

**Comment by yuwata** [comment]
Should be duplicate of #24150.

**Comment by yuwata** [comment]
Fix is waiting in #31311.
